tensorboard/data/server/reservoir.rs

```diff
@@ -234,6 +236,53 @@ impl<T, C: ReservoirControl> StageReservoir<T, C> {
         &self.staged_items[..]
     }
 
+    /// Checks whether a preemption has been staged.
+    ///
+    /// This is initially `false`; it changes to `true` whenever a record is preempted, and back to
+    /// `false` whenever the reservoir is committed.
+    pub fn staged_preemption(&self) -> bool {
+        self.staged_preemption
+    }
+
+    /// Preempts any records whose step does not precede the given step.
+    fn preempt(&mut self, step: Step) {
+        let staged_nonpreempted = self
+            .staged_items
+            .iter()
+            .enumerate()
+            .rev()
+            .find_map(|(i, (s, _))| if *s < step { Some(i + 1) } else { None })
```

Comment:
I struggled a little to grok the iterator chain here - some of that is my having to look up APIs like `find_map()` but then it also just felt a bit convoluted to have this predicate checking for non-preempted items in order to count those (but not by directly counting them per se; instead by returning the enumerated index), all in order to subtract those from the length and then truncate accordingly.

Is there a reason not to just use `binary_search_by_key()` which looks like it would directly return the truncation point?  I'm not sure if I'm missing some reason that doesn't work, or if there's a performance concern about binary search or something else.

If we did want to linearly scan from the end of the vector, having looked at the Iter APIs for a while, what seems more "natural" to me would be:

```rust
let staged_preempted = self.staged_items.iter().rev().take_while(|(s, _)| s >= step).count();
```

Is there a reason the version you have is preferable?  (Honest question.)

Reply:
Yeah I think it's the "citation needed" part where I'm a bit unsure.  I agree it's a bit more obvious to think of preempting only a few trailing events, but in practice I think I've also seen jobs that preempt back to 0 several times (think of a job failing repeatedly to even get to the first checkpoint) so I figured binary search avoids the worst case, especially for large reservoirs.  But I'm 100% happy to wait until further along and benchmark against some actual data.

Reply:
(And yes, it is very cool that the compiler can optimize through all those layers!)